DESIGN.md

```diff
@@ -0,0 +1,156 @@
+We could make `nupm` itself an overlay that adds itself as a command.
+
+There are several approaches:
+* bootstrap using shell script sourced from web (like rustup)
+* embedded inside Nushell's binary
+  * The advantage of this is that it does not require user's config. The PATH and NU_LIB_DIRS could be pre-configured in Nushel
+* (in the future maybe) as a compiled binary (using something like https://github.com/jntrnr/nu_app)
+  * This would allow us to reverse the installation steps: Instead of Nushell installing nupm, we could let user only install nupm which would in turn install Nushell
+
+## Dependency handling
+
+In compiled programming languages, there are two kinds of dependencies: static and dynamic. Static are included statically and compiled when compiling the project,
+dynamic are pre-compiled libraries linked to the project.
+Note that Nushell is [similar to compiled languages][Nushell compiled] rather than typical dynamic languages like Python, so these concepts are relevan for Nushell.
+
+Static dependencies:
+* Advantages: reproducible, does not rely on system files (no more missing random.so.2), higher performance (allows joint optimization of dependencies and project itself)
+* Disadvatage: increased compile time, binary size, can easily end up with multiple versions of the same library (hello Nushell dependencies)
+
+Dynamic dependencies are the opposite basically. Note that Nushell currently supports only static dependencies, but we might be able to add the "linking" feature at some point.
+
+We might want `nupm`support both types of dependencies.
+
+## Package repository 
+
+Packages need to be stored somewhere. There should be one central "official" location (see https://github.com/NixOS/nixpkgs for inspiration).
+
+Additionally, user should be able to add 3rd party repositories as well as install local and other packages (e.g., from the web, just pointing at URL),
+as long as it has `METADATA_FILE` telling `nupm` what to do.
+
+## API / CLI Interface
+
+Nushell's module design conflates CLI interface with API -- they are the same.
+
+- `nupm new [--script] [--module]`
+    - create a new local package with template files ([`kubouch/nuun`])
+- `nupm list`
+    - list currently installed packages and if they're out of date
+- `nupm install`
+    - install package into the currently active overlay (can override which overlay to install to)
+- `nupm add`
+    - add a dependency to the current project
+    - it is different from `nupm install`: this one adds the dependency to the MANIFEST_FILE, `nupm install` does not
+- `nupm uninstall`
+    - uninstall a package from a currently active overlay (can override which overlay to install to)
+- `nupm update`
+    - update all packages in a currently active overlay overlay (can specify package and/or overlay name)
+    - can be used to self-update (let's avoid having both "update" and "upgrade")
```

Comment:
Agreed, though I would guard `nupm` auto-update with `--self-upgrade` or sth like that, so it's not a default behavior. One should be able to prevent `nupm` from surprise updates that can break their all workflow

Reply:
i have
- `nupm update nupm`
- `nupm update --self`
- `nupm update --all`

in my `amtoine/nupm` that would all update `nupm` itself :+1:

Reply:
Or `nupm update self`, agreed.

Reply:
> Or `nupm update self`, agreed.

I would rather see it as a flag than a separate subcommand tbh. @amtoine's suggestion looks perfect t me 😋

Reply:
done in 6fcebf82c2636bf68e9cecee613a4f8af497f221